rpc: introduce 'label' API for wallet

laanwj opened the pull request:
Add label API to wallet RPC.

This is one step towards #3816 ("Remove bolt-on account system") although it doesn't actually remove anything yet (that would be a follow-up pull).

These initially mirror the account functions, with the following differences:
- These functions aren't DEPRECATED in the help
- Help mentions 'label' instead of accounts. In the language used, labels are associated with addresses, instead of addresses associated with labels. (in contrast to accounts.)
- Labels have no balance
  - No `getreceivedbylabel`
  - No balances in `listlabels`
  - `listreceivedbylabel` can show received transactions to addresses with a label, not use the account tally (currently it is removed, but according to discussion that goes too far as this doesn't inherently have to do with balance)
  - `listlabels` has no minconf or watchonly argument
  - Remove `move`
- Like in the GUI, labels can be set on any address, not just receiving addreses
- Unlike accounts, labels can be deleted. Being unable to delete them is a common annoyance (see #1231).  Currently only by reassigning all addresses using `setlabel`, but an explicit call `deletelabel` which assigns all address to the default label may make sense.
- These calls stay the same, with `account` argument renamed to `label`:
  - `importaddress`
  - `importprivkey`
  - `importpubkey`

## API

Short description of every RPC call: for detailed information check RPC help. The general idea is to offer the same functionality as the GUI label system. Labels are simply a name for an address, or a group of addresses.

**Do not use the deprecated account system and the label system with the same wallet at the same time**. These APIs use the same underlying data in the database for (slightly) different purposes, using them interchangeably will give unexpected results. (Just like using the GUI labels and account system at the same time. Using the GUI labels and the label API at the same time, however, is no problem)
- `getlabel`: returns the label (and other address book data) associated with an address
  - This exposes the fields in the CAddressBookData structure, which is currently the 'purpose' (sending address, receiving address) and 'destdata' (used for storing payment requests IIRC)
- `getaddressesbylabel`: get addresses labelled with one label
- `listlabels`: list all labels (or labels with a certain purpose, such as receive/send)
- `setlabel`: assign a label to an address
- `getlabeladdress`: get the 'label address' for the specified label. This gets an unused address with the label, creating one if necessary **should be removed according to discussion**

These calls have a deprecated account parameter, which can be turned into a label-parameter as is:
- `listtransactions`

## Open questions

- Should there be such a thing as a 'label address'? My initial feeling about this was
  'no', labels are just a name for one or more addresses, intuitively there is no "default address",
  and it also isn't a GUI feature (->no)

laanwj committed:
rpc: introduce 'label' API for wallet

Add label API to wallet RPC.

This is one step towards #3816 ("Remove bolt-on account system") although it doesn't
actually remove anything yet.

These initially mirror the account functions, with the following differences:

- These functions aren't DEPRECATED in the help
- Help mentions 'label' instead of accounts. In the language used, labels are
  associated with addresses, instead of addresses associated with labels. (unlike
  with accounts.)
- Labels have no balance
  - No balances in `listlabels`
  - `listlabels` has no minconf or watchonly argument
- Like in the GUI, labels can be set on any address, not just receiving addreses
- Unlike accounts, labels can be deleted.
  Being unable to delete them is a common annoyance (see #1231).
  Currently only by reassigning all addresses using `setlabel`, but an explicit
  call `deletelabel` which assigns all address to the default label may make
  sense.

jonasschnelli commented:
Concept ACK.
I just wonder if this makes the wallet code more complex (add another layer).

My idea was it to duplicate the current wallet implementation (`cp src/wallet src/newwallet`-approach) and add such things there (after removing the accounts-related code). Also the Bip32 and @CodeShark multi-wallet PR could be added there.

The second wallet could come without API stableness (for the first two releases or so) and could be marked as experimental.

laanwj commented:
> I just wonder if this makes the wallet code more complex (add another layer).

I disagree:

- All of this functionality is required anyway to support the GUI.
- Yes, there is some intentional duplication, but only until the account calls are ripped out, which should be one of the next steps.
- There are only added RPC calls in rpcwallet.cpp. The `CWallet` class is not complicated by this.

The point here is to give a non-deprecated replacement for the 'label' system as used in the GUI, so the subset that people are still allowed to use.

jonasschnelli commented:
> As for alternative wallets, they've been proposed since at least 2012 - but none have materialized yet. And none of this change rules them out.

Yes. I agree.
This PR has a clear value.

luke-jr commented:
Note getaccountaddress does not presently get a "default" address, it gets an *unused* address with the label, creating one if necessary. This seems useful *only* for mining, since no other context can guarantee an address hasn't been "used" but not sent to yet. I can't think of a good way to deprecate this, however.

laanwj commented:
> it gets an unused address with the label, creating one if necessary.

I wonder if we can find a better (or at least simpler) way to do this, now that we're creating a new API anyway.

laanwj commented:
@luke-jr 
> I wonder if we can find a better (or at least simpler, the GetAccountAddress function is pretty terrible) way to do this, now that we're creating a new API anyway. Need to think about this a bit.

I thought of the following: you could use two labels, one for the 'active' address, one for the 'normal'. Say `mining_active` and `mining`.

When the miner needs an address it will:
- *addr* = `getaddressesbylabel mining_active`
- `getreceivedbyaddress` - check with wallet that *addr* has been used before (or maybe add a more convenient RPC call for checking a single address)
- If so:
  -  `setlabel <addr> mining`  - move current address to normal label
  - *addr* = `getnewaddress mining_active` - generate new address in `mining_active` label
  - Use *addr* for mining to
- If not:
  - Use *addr* for mining to

This is a little bit more complicated at the user side, but it avoids special administration at the server side.

luke-jr commented:
That looks like a lot of overhead, and this is a rather time-sensitive call, as the miner is working on stale work until it's done.

Also, why are there no getreceivedbylabel/listreceivedbylabel? These don't have anything to do with balances.

laanwj commented:
> That looks like a lot of overhead, and this is a rather time-sensitive call, as the miner is working on stale work until it's done.

I'd suggest to try it. It shouldn't be much slower.

> Also, why are there no getreceivedbylabel/listreceivedbylabel? These don't have anything to do with balances.

Looks like you're right. `getreceivedbyaccount` doesn't actually return the account balance, but the total number of coins sent to the addresses that make up the account?

`listreceivedbyaccount`  on the other hand goes over the account tally. But I agree if it only were to show actual transactions sent to addresses belonging to a label it'd be ok.

sipa commented:
@luke-jr Wouldn't it be feasible to instead generate a sequence of deterministic addresses for mining, for example using BIP32 derivation with the block height as index?

luke-jr commented:
> I'd suggest to try it. It shouldn't be much slower.

@laanwj getreceivedbyaddress at least would loop over all the wtx... and then there's the additional latency from the back and forth of multiple calls. I haven't tried it yet, though.

> Wouldn't it be feasible to instead generate a sequence of deterministic addresses for mining, for example using BIP32 derivation with the block height as index?

@sipa Perhaps, if the wallet had a way to do this. Using the height seems incompatible with gap limits, though?

sipa commented:
@luke-jr I mean the mining software can do derivation, and import the keys into the wallet when a block is found.

MarcoFalke commented:
The functional test coverage for accounts is minimal or not existent, I think we should move forward with this pull.

Needs rebase.

jonasschnelli commented:
Re-Concept ACK.
I think this solution makes more sense than the closed #7830.
Needs tests, rebase and release-note mentioning.

wallclockbuilder commented:
Needs tests.

laanwj commented:
@wallclockbuilder No shit, have you seen the TODOs at the bottom of the opening post?

laanwj commented:
To be clear I posted this to get comments on the API, is there anything left to be done here? I'm going to write tests when it is clear that this is what we want at all.

MarcoFalke commented:
> Do not use the deprecated account system and the label system with the same wallet at the same time
> [...]
> optional: a flag in the wallet to prevent use of both the account and label API

I would not consider this optional. User will always do what you not want them to do.

sipa commented:
Concept ACK

andrewbaine commented:
listtransactions has an "account" where you now you would pass "*" if you need to supply non-default args for count, from, and includeWatchOnly. Will there be a way to query for transactions affecting any address with a given label? Could we tack on an argument to listtransactions?

laanwj commented:
I think the account argument of listtransactions could simply be re-used as a label argument. As listing transactions to a label has nothing to do with per-label balances there is no need to drop that particular functionality,

andrewbaine commented:
repurposing the "account" argument to be "label" makes sense to me

jtimon commented:
Fast review ACK (besides needing rebase and the promised tests).
The API is actually more than I expected but still simple enough, I think.

 > optional: a flag in the wallet to prevent use of both the account and label API

I'm not sure it's worth to bother. I think we should go ahead and completely remove account functionality within 0.14 instead. But whatever we do, it can be done in a later PR.

instagibbs commented:
>labels are associated with addresses, instead of addresses associated with labels
Description nit: not sure I catch the distinction here.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
     return result;
 }
 
+UniValue getlabeladdress(const UniValue& params, bool fHelp)
```
would `getnewlabeladdress` be wrong? As-is gave me the impression that this was some static address.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+            "\nArguments:\n"
+            "1. \"label\"       (string, required) The label for the address. It can also be set to the empty string \"\" to represent the default label.\n"
+            "\nResult:\n"
+            "\"bitcoinaddress\"   (string) The 'label address' for the label\n"
```
perhaps mention `latest unused` or something similar.

laanwj replied:
I still think we should get rid of this.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+            "listlabels ( \"purpose\" )\n"
+            "\nReturns the list of all labels, or labels that are assigned to addresses with a specific purpose.\n"
+            "\nArguments:\n"
+            "1. \"purpose\"  (string, optional) Address purpose to list labels for ('send','receive'). An empty string is the same as not providing this argument.\n"
```
I'm not as familiar with accounts as others, but is this type of parameter actually going to be used? When would you want to list labels that go certain directions? I have a feeling labels people choose will reflect this already.

laanwj replied:
The goal of this API is to expose exactly the same functionality that the GUI uses. The GUI distinguishes between different kind of labels (to show sending/receiving addresses), so it should be offered here as well.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    if (fHelp || params.size() < 1 || params.size() > 2)
+        throw runtime_error(
+            "setlabel \"bitcoinaddress\" \"label\"\n"
+            "\nSets the label associated with the given address.\n"
```
it also adds it to the address book if previously unknown, correct?

laanwj replied:
Indeed, that is what setting a label does. If there is no labeling information associated with an address it will create that. We don't want to use the term "address book" here I think.

instagibbs reviewed (commented):
concept ACK, with some nits/questions.

I am not an expert on how accounts exactly work as-is.

It also may be helpful to give a brief motivation in the OP. What's wrong with accounts, and what label fixes about that. Currently it's a list of differences without clear motivation.

laanwj commented:
> It also may be helpful to give a brief motivation in the OP. What's wrong with accounts, and what label fixes about that. Currently it's a list of differences without clear motivation.

See #3816

jonasschnelli commented:
utACK a2557ffb80543af5e92c3db03d257fe594b0d910 needs (trivial) rebase for the new `JSONRPCRequest` object passing.

instagibbs commented:
@laanwj yes I read the issue, but there are some disagreements in that thread about what the actual issue is. I assume it's along the lines of "people want to watermark addresses, but bitcoind wallet shouldn't try to be an accounting system for those labels".

motatoes commented:
Hi. When is this new feature expected to be rolled out?

sipa commented:
@motatoes Like everything, when it's ready. That may be in 0.14.0 or later.

ryanofsky commented:
@laanwj , I created an RPC test for this change here: https://github.com/ryanofsky/bitcoin/commit/2dac8eb0d915b4c04764eff100d81718b2cd9a90

Feel free to incorporate it in this PR, or I could create a separate one.

jtimon commented:
Needs rebase

Tintress15 reviewed (approved):
Get a balance

jtimon commented:
Needs rebase again, sorry for not reviewing after the last rebase.

sipa commented:
Concept reACK

TheBlueMatt reviewed (commented):
Grr, posted on wrong PR...(and cant delete reviews?)

TheBlueMatt commented:
Needs rebase. Concept ACK

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+            "\nSets the label associated with the given address.\n"
+            "\nArguments:\n"
+            "1. \"bitcoinaddress\"  (string, required) The bitcoin address to be associated with an label.\n"
+            "2. \"label\"           (string, required) The label to assign to the address.\n"
```
Is the label optional or required?

If it's required update L2675 to `if (fHelp || params.size() != 2)` and L2693-2695 to `std::string strLabel = AccountFromValue(params[1]);`, if optional update the docstring.

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+        pwalletMain->SetAddressBook(address.Get(), strLabel, "receive");
+    }
+    else
+        pwalletMain->SetAddressBook(address.Get(), strLabel, "send");
```
Braces please!

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2537,13 +2537,13 @@ UniValue getlabeladdress(const UniValue& params, bool fHelp)
 {
     UniValue ret(UniValue::VOBJ);
-    if (includeName)
+    if (verbose)
```
Does the order of push_back matter here? Might be cleaner to lump all `if (verbose)` under one branch...

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    UniValue ret(UniValue::VSTR);
+
+    ret = GetAccountAddress(strLabel).ToString();
+    return ret;
```
It seems like L2529-2534 could be a one or two liner, rather than 4 (and maybe get rid of the whitespace).

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    ret.push_back(Pair("purpose", data.purpose));
+    if (includeDestData) {
+        UniValue ddata(UniValue::VOBJ);
+        BOOST_FOREACH(const PAIRTYPE(std::string, std::string) &item, data.destdata)
```
does a builtin for range loop work here?

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+
+    // Find all addresses that have the given label
+    UniValue ret(UniValue::VOBJ);
+    BOOST_FOREACH(const PAIRTYPE(CBitcoinAddress, CAddressBookData)& item, pwalletMain->mapAddressBook)
```
Can this not be a for range loop?

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+        purpose = params[0].get_str();
+
+    std::set<std::string> setLabels;
+    BOOST_FOREACH(const PAIRTYPE(CTxDestination, CAddressBookData)& entry, pwalletMain->mapAddressBook) {
```
can use for range?

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+            setLabels.insert(entry.second.name);
+    }
+    UniValue ret(UniValue::VARR);
+    BOOST_FOREACH(const std::string &name, setLabels)
```
can use for range?

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+
+    std::string purpose;
+    if (params.size() > 0)
+        purpose = params[0].get_str();
```
validate the purpose string?

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+
+    if (IsMine(*pwalletMain, address.Get()))
+    {
+        // Detect when changing the label of an address that is the 'label address' of another label:
```
This behavior is probably just to mirror prior behavior, but perhaps a better alternative would be to create a fresh address for the account & allow deleting of account via another mechanism.

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2497,6 +2503,216 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    if (params.size() > 1)
+        strLabel = AccountFromValue(params[1]);
+
+    if (IsMine(*pwalletMain, address.Get()))
```
What's the correct behavior when address is watchonly?

JeremyRubin reviewed (approved):
utack!

Overall, looks good. Just nits from me really. Only thing I was concerned about is the proper behavior for watchonly addresses, perhaps some added documentation for that would be good.

ryanofsky commented:
@laanwj, do you want to cherry pick my unit test into this PR? (https://github.com/ryanofsky/bitcoin/commit/2dac8eb0d915b4c04764eff100d81718b2cd9a90)

In [IRC recently](https://botbot.me/freenode/bitcoin-core-dev/msg/83565594/), you requested that people "please review the *API*, not the code," and I think the unit test can help with this, because it demonstrates the API in action, covers various subtleties, and is also well commented.

laanwj commented:
> @laanwj, do you want to cherry pick my unit test into this PR? (ryanofsky@2dac8eb)

Yes, I will, thank you!

TheBlueMatt commented:
API ACK. Still needs rebase :(

fiubit commented:
[http://double-bit.site123.me/](url)
i want her ! Core hallo

gmaxwell commented:
I expected to be able to attach multiple labels to a single address.

laanwj commented:
> I expected to be able to attach multiple labels to a single address.

That is not possible in the GUI either, in the OP I've defined the scope as:

> The general idea is to offer the same functionality as the GUI label system

Such functionality could be added in the future.

jtimon commented:
Nobody seems to have big problems with the API for a while. Perhaps it's time to rebase and start reviewing the code itself

kewde commented:
@laanwj
https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-200219840

>addr = getnewaddress mining_active - generate new address in mining_active label
>Use addr for mining to

I'm not completely aware of what exactly happens in getnewaddress, but [documentation](https://bitcoin.org/en/developer-reference#getnewaddress) seems to state that the argument 'mining_active' would be the "account name" and not the label.

jnewbery commented:
I wanted to test this so I rebased on master and cherry-picked Russ's test. I've pushed it here: https://github.com/jnewbery/bitcoin/tree/pr7729 . Feel free to reset onto that commit. Note that there were quite a few refactors (multiwallet, namespaces, database wrapper, JSONRPC request, rpc names args), so someone else should probably review the rebase to make sure I haven't missed anything.

Note that the final test in Russ's test script currently fails. I added some trace to the script and see that the address returned by `getlabeladdress` isn't included in the list of addresses returned by `getaddressesbylabel`:

```
-146-> getaddressesbylabel ["c"]
<-146- [0.000135] {"n1HsWg6ANV5Z8SLojKNyjqCRf5saqQKQed": {"purpose": "receive"}, "mfYMvaXrHp2RUxe1q1PwRgS8AokYjgUSmf": {"purpose": "receive"}, "mhVHLa9KCHGMXHBNpyfAHNZXmwKaaTg6Fj": {"purpose": "receive"}}
-147-> getlabeladdress ["c"]
<-147- [0.001559] "mnvNe2eG3hmderC3y5E4SYLNnwWsYrSg1c"
-148-> getlabel ["mnvNe2eG3hmderC3y5E4SYLNnwWsYrSg1c"]
<-148- [0.000136] {"name": "c", "destdata": {}, "purpose": "receive"}
```

I haven't yet dug into why that's the case.

jnewbery committed:
[wallet] Make setlabel idempotent

Prevent setlabel calls on an address that already has the same label, and which
is the already the default "label address" of that label, from generating a
new default address.

This way calling setlabel on an address that already has the specified label is
a no-op.

Change suggested by Russell Yanofsky <russ@yanofsky.org> in
https://github.com/bitcoin/bitcoin/pull/7729#discussion_r92064042

jnewbery commented:
I've made a few more changes to rpcwallet.cpp to reflect more recent style guidelines (no `BOOST_FOREACH` and braces for if statements. I've also committed a fixup commit for the failing test case. @laanwj feel free to squash that if you're happy with it.

Once you've taken the changes I'll add my review comments.

laanwj commented:
@jnewbery Thanks so much! Updated this pull to that branch.

jnewbery commented:
needs rebase again! (although this one should be easier - just fixing the `BASE_SCRIPTS` list in test_runner.py)

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2923,6 +2928,236 @@ UniValue bumpfee(const JSONRPCRequest& request)
+    LOCK2(cs_main, pwallet->cs_wallet);
+
+    // Parse the label first so we don't generate a key if there's an error
+    std::string strLabel = AccountFromValue(request.params[0]);
```
It'd be nice to eventually rename this function `LabelFromValue()`. That can be done in a follow-up PR.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2923,6 +2928,236 @@ UniValue bumpfee(const JSONRPCRequest& request)
+            "listlabels ( \"purpose\" )\n"
+            "\nReturns the list of all labels, or labels that are assigned to addresses with a specific purpose.\n"
+            "\nArguments:\n"
+            "1. \"purpose\"  (string, optional) Address purpose to list labels for ('send','receive'). An empty string is the same as not providing this argument.\n"
```
nit: I don't think `An empty string is the same as not providing this argument.` is required

promag replied:
Agree with @jnewbery, empty value should be an error then.

promag replied:
BTW, why is this argument needed? I mean, the client can filter the result. IMO pagination is more interesting.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2923,6 +2928,236 @@ UniValue bumpfee(const JSONRPCRequest& request)
+
+    if (IsMine(*pwallet, address.Get()))
+    {
+        // Detect when changing the label of an address that is the 'label address' of another label:
```
I think we should remove the concept of 'label address' and change this section to delete the label when removing the final address from that label.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2923,6 +2928,236 @@ UniValue bumpfee(const JSONRPCRequest& request)
+    }
+
+    std::string strLabel;
+    if (request.params.size() > 1){
```
What happens if `request.params.size() == 1`? Are you supposed to be able to remove a label using this rpc without a second argument? If so, I think you want to call `pwallet->DelAddressBook`.

MeshCollider replied:
Is there any difference between setting an empty label vs deleting the label?

laanwj replied:
Well if you set all addresses with a certain label to the empty label, that label is effectively deleted. Deleting the label is a shortcut to remove it from all addresses.

jnewbery reviewed (commented):
A few comments inline. More general feedback:

1. I really don't like the concept of a 'label address'. It seems to go against the philosophy of 'labels are associated with addresses, instead of addresses associated with labels', and makes the distinction between labels and accounts more confusing.
You also mentioned in the OP *this would greatly simplify the code though - it would be impossible to get rid of the CAccount structure due to this.*.
So it sounds to me like removing label addresses makes the interface clearer for users and makes the implementation simpler, so let's just do it. https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-200219840 seems to be a sensible solution for miners using account addresses.

2. I agree that users should be able to add multiple labels to an address. That can be done as a follow-up PR.

3. You say in the OP **Do not use the deprecated account system and the label system with the same wallet at the same time.**. There doesn't appear to be any discussion yet on how to achieve this, or how to handle migrating from accounts to labels. One suggestion: hide these new RPCs with old wallet versions, add a new `enablelabels` RPC which upgrades the wallet to a new version, and on the new version, hide the account RPCs. The `enablelabels` RPC should describe the implications of moving from accounts to labels. I think this requires a bit more thought and discussion.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3146,7 +3146,7 @@ UniValue setlabel(const JSONRPCRequest& request)
         {
             std::string strOldLabel = pwallet->mapAddressBook[address.Get()].name;
-            if (address == GetAccountAddress(pwallet, strOldLabel)) {
+            if (strOldLabel != strLabel && address == GetAccountAddress(pwallet, strOldLabel)) {
```
Note: This change makes calling setlabel on an address which already has the same label a no-op, see https://github.com/bitcoin/bitcoin/pull/7729#discussion_r92064042.

laanwj commented:
So to summarize above discussion, the following changes should be made:
- Add `listreceivedbylabel`, listing transactions to addresses that have a certain label
- Remove the default label address. Discussion confirms my intuition that it's better to get rid of this. 
  - 'acc' database entries, as well as CAccount, would not be necessary for labels. After this change, labels have no metadata, so a lot of code can be simplified.
- `setlabel` on an address that already has the same label should be a no-op
- Documentation changes / nits

>  RPC which upgrades the wallet to a new version, and on the new version, hide the account RPCs. The enablelabels RPC should describe the implications of moving from accounts to labels. I think this requires a bit more thought and discussion.

Agree re: versioning. We already have a command line argument to upgrade the wallet version, `-upgradewallet`. Wouldn't that be enough?
Either label or account RPCs could be usable based on the wallet version.
I don't see why a RPC to do the upgrade would be preferable, it's something that will be done at most once.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2922,6 +2927,236 @@ UniValue bumpfee(const JSONRPCRequest& request)
+    }
+    ret.push_back(Pair("purpose", data.purpose));
+    if (verbose) {
+        UniValue ddata(UniValue::VOBJ);
```
Nit, rename `ddata` to `dest_data`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -152,6 +152,11 @@ UniValue getnewaddress(const JSONRPCRequest& request)
     return CBitcoinAddress(keyID).ToString();
 }
 
+void DeleteAccount(CWallet * const pwallet, std::string strAccount)
```
Nit, shouldn't we follow the convention in new code in favor of consistency? In that case rename `strAccount` to `account_name` for instance?

promag reviewed (commented):
Partial review. Is this going forward?

jnewbery commented:
> We already have a command line argument to upgrade the wallet version, -upgradewallet. Wouldn't that be enough

ACK - yes this is enough.

The wallet startup command line arguments always seemed strange to me, and if we were building this from scratch we might implement upgradewallet as an RPC or an external tool. However, it makes sense to use the existing infrastructure.

> Either label or account RPCs could be usable based on the wallet version.

✅ sounds good to me.

Two additional points:

1. Obviously documentation is key here. The wallet upgrade will effectively remove the accounts from the wallet, so we need to make sure users understand the implication before upgrading.
2. When will we be able to remove the account code entirely? At that point will bitcoind no longer support wallets older than the new version? Is there precedence for how we deprecate wallet features like this?

> Is this going forward?

It feels like we've missed the boat for 0.15. @laanwj - it'd be great to get this in early in the next cycle. I'll have some spare time in the next few weeks. Let me know if there's anything I can do to help here. I'm happy to help write and test the upgrade code.

mmgen commented:
@laanwj `importaddress`, `importpubkey` and `importprivkey` have "label" arguments that currently set the account for an address (deleting the old one).  Won't these calls need to be updated as well? I don't see them in your patch.

mmgen commented:
@jnewbery Since I got no reponse from @laanwj, I'll refer you to my above comment. The behavior I mentioned can easily be demonstrated with the following:

	$ bitcoin-cli importaddress n2FgXPKwuFkCXF946EnoxWJDWF2VwQ6q8J 'Label 1'
	$ bitcoin-cli listaccounts 0 true | grep 'Label 1'
	 "Label 1": 0.00000000,
	$ bitcoin-cli importaddress n2FgXPKwuFkCXF946EnoxWJDWF2VwQ6q8J 'Label 2'
	$ bitcoin-cli listaccounts 0 true | grep 'Label 1'
	$ bitcoin-cli listaccounts 0 true | grep 'Label 2'
	 "Label 2": 0.00000000,

jnewbery commented:
@mmgen : @laanwj is very busy with the 0.15 release right now, so I'm not surprised he hasn't updated this PR. I'm planning on picking this up early in 0.16 and I'll review your comments then. Any help with review/testing then will be very much appreciated!

mmgen commented:
@jnewbery Thanks! Will be ready to help when needed.

achow101 commented:
Concept ACK, needs rebase.

Maybe this and #11497 should be done together?

Earlier in this thread, it was suggested that the version number be bumped to distinguish between whether to use accounts or wallets. Because of the HD wallets thing with the optional features in the version number, would we want to bump the version number (which would then mean only HD wallets can use labels) or do something else?

ryanofsky commented:
Should probably just drop the version number idea now that we have the `-deprecatedrpc` mechanism. Seems like adding versioning to the decision of which rpcs to enable would make the transition more complicated without providing a clear benefit.

MeshCollider commented:
I've rebased this on current master here: https://github.com/MeshCollider/bitcoin/tree/201710_labels_rebased

Todo summary from comments:
- Remove `getlabeladdress` RPC
- Add `listreceivedbylabel` RPC
- Incorporate #11497 to disable accounts at the same time (including renaming of `account` params to `label`)

mmgen commented:
`importaddress`, `importpubkey` and `importprivkey` have 'label' arguments (see my note above) that currently set the account for the address. After the label API is introduced, will these now set the label, or the account? Since both APIs will continue to exist side-by-side for a time, this could be a source of potential confusion for users. My wallet uses `importaddress` to set the account, which I use as an alias for a given address.

laanwj commented:
> importaddress, importpubkey and importprivkey have 'label' arguments (see my note above) that currently set the account for the address. After the label API is introduced, will these now set the label, or the account?

The same arguments will be used for label (where it makes sense) to ease the transition and avoid unused holes in the argument list. This is mentioned in the OP.
For 0.16 it will be possible to switch between label and account API with a command line argument, but will default to labels. 0.17 will likely drop suppport for accounts completely.

ryanofsky commented:
Rebased this on top of #11536 in 1dbfe49c48077ee752c6851b411f89ccf679cb2e (https://github.com/ryanofsky/bitcoin/commits/pr/label)

sipa commented:
Can we haz this for 0.16?

ryanofsky commented:
> Can we haz this for 0.16?

I think the best way to move this forward would be to review and merge #11536, which gets account -> label renaming out of the way so this PR can be small and simple and doesn't need to be constantly rebased.

laanwj commented:
0.17, not 0.16. 0.16 is supposed to be the intermediate segwit wallet
release, I'd strongly prefer to not add any other features for it that
potentially hold it up longer.

On Jan 6, 2018 11:51, "Pieter Wuille" <notifications@github.com> wrote:

> Can we haz this for 0.16?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-355739097>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHutpkOslEiyC81Ta-E4I1OTix7NPwBks5tH1A8gaJpZM4H1S2Y>
> .
>

Sjors commented:
The rebased version seems quite different from this, yet is relevant in order to understand #11536. Maybe this PR should be closed and replaced?

ryanofsky commented:
> The rebased version seems quite different from this (or at least from the description), yet is relevant in order to understand #11536. Maybe this PR should be closed and replaced?

This should not be true. The rebased version Iinked in my comment above https://github.com/bitcoin/bitcoin/pull/7729#issuecomment-338417139, has the exact same changes as the current version of this PR, plus some additional "account" -> "label" for additional "account" occurrences missed here. #11536 + #7729 should be a pure superset of #7729.

If there are any actual differences that you see, please point them out in #11536 or here so I can fix my branch.

Sjors commented:
Maybe it's only the description of this PR that's out of date. In particular this list suggests that label RPC commands are a subset of the original account commands:
 
<img width="675" alt="schermafbeelding 2018-01-18 om 10 10 48" src="https://user-images.githubusercontent.com/10217/35089669-ec82bf90-fc37-11e7-8927-e651faba2364.png">

Whereas the actual code and commit messages don't remove any methods.

ryanofsky commented:
The description at the top of this PR is slightly out of date. It should be edited to match the description in the actual commit message, which should be fully up to date.

> Whereas the actual code and commit messages don't remove any methods.

I think you are taking "remove" in the description too literally. Perhaps it could be rephrased. This PR is adding label functions, not removing existing account functions (that meant to be done later for to preserve compatibility). When the description says "remove" it is just referring to not adding a new label function corresponding to an existing account function.

Sjors commented:
Can you add a remark at the top that this depends on #11536?

ryanofsky commented:
Rebased PR at https://github.com/ryanofsky/bitcoin/commits/pr/label:

- 6a0d27412d9c1b16e4bdfc406ee1e7b0ee6a2a51 rpc: introduce 'label' API for wallet
- fef41789e8d7293ccfd859bc90fa0a3b2484752d [wallet] Make setlabel idempotent

laanwj commented:
@ryanofsky Thank you! Replaced the branch with that one.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -290,14 +295,14 @@ UniValue setlabel(const JSONRPCRequest& request)
             "\nExamples:\n"
-            + HelpExampleCli("setlabel", "\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\" \"tabby\"")
-            + HelpExampleRpc("setlabel", "\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XX\", \"tabby\"")
+            + HelpExampleCli("setlabel", "\"1D1ZrZNe3JUo7ZycKEYQQiQAWd9y54F4XZ\" \"tabby\"")
```
Unrelated change?

If you really wanted to change that `DummyAddress(Params())` would be a better choice.

Sjors commented on src/wallet/rpcwallet.cpp:
```diff
@@ -208,11 +213,11 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
-            "\nReturns the current Bitcoin address for receiving payments to this label.\n"
+            "\nReturns the current 'label address' for this label.\n"
             "\nArguments:\n"
-            "1. \"label\"         (string, required) The label name for the address. It can also be set to the empty string \"\" to represent the default label. The label does not need to exist, it will be created and a new address created  if there is no label by the given name.\n"
```
I'm fine with removing this remark, but note that the behavior of creating a new address still exists (though without an address type param). It might make sense to deprecate that behavior in a followup PR.

Sjors reviewed (approved):
tACK fef41789e8.

I just noticed that changing or removing a label in the console live updates the Receiving Addresses UI, nice! For a future PR we should figure out what to do with the Receive and Transaction tab "labels", as these seem to use an independent mechanism.

laanwj commented:
> For a future PR we should figure out what to do with the Receive and Transaction tab "labels", as these seem to use an independent mechanism.

Transaction tab labels use the same mechanism so it would be a matter of listening to a notification and repainting. Receive requests are stored separately, because they have extra metadata besides a label.

Sjors commented:
> Transaction tab labels use the same mechanism so it would be a matter of listening to a notification and repainting.

Even restarting QT doesn't change or remove a label from Requested payments history for me.

Sjors commented:
Ah yes, I see it. Restarting QT doesn't change or remove a label from Requested payments history for me, as you point out. Switching tabs is enough to update labels in the transactions view.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -208,11 +213,11 @@ UniValue getlabeladdress(const JSONRPCRequest& request)
         throw std::runtime_error(
             "getlabeladdress \"label\"\n"
-            "\nReturns the current Bitcoin address for receiving payments to this label.\n"
+            "\nReturns the current 'label address' for this label.\n"
```
The behaviour for this RPC is weird. If called for a label that doesn't exist, it creates a new label, and then adds a new address as the 'label address' for that label. That's not very intuitive, and I think it's a bad experience (for example if a user typos an existing label name). Can we change this so that the rpc returns an error if called with a non-existent label name?

jnewbery commented on test/functional/wallet_labels.py:
```diff
@@ -15,6 +15,7 @@
 
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import assert_equal
+from collections import defaultdict
```
nit: standard library imports before project imports please!

jnewbery reviewed (commented):
Comments inline.

Feel free to squash my second commit into the first.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -290,14 +295,14 @@ UniValue setlabel(const JSONRPCRequest& request)
     if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
         throw std::runtime_error(
-            "setlabel \"address\" \"label\"\n"
+            "setlabel \"bitcoinaddress\" \"label\"\n"
```
> Why rename the arguments/return values to bitcoinaddress everywhere? 

Agree it would be better to stick to `address`, but just as a historical note, this wasn't a "gratuitious" API break when it was originally written in 8571fee617aa22a43d73efa2560cd2ddf55c2478, because it preceded #11536, so the `setlabel` RPC was brand new.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3809,6 +3816,152 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
+            "\nArguments:\n"
+            "1. \"bitcoinaddress\"  (string, required) The bitcoin address for label lookup.\n"
+            "\nResult:\n"
+            "  { (json object with information about address)\n"
```
> Can we make this RPC return an array?

It might be better to not add a `getlabel` RPC at all but instead just return this information in `getaddressinfo` (recently added in #10583).

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3882,6 +4024,26 @@ static const CRPCCommand commands[] =
+    { "wallet",             "getaccountaddress",                &getlabeladdress,               {"account"} },
+    { "wallet",             "getaccount",                       &getaccount,                    {"address"} },
+    { "wallet",             "getaddressesbyaccount",            &getaddressesbyaccount,         {"account"} },
+    { "wallet",             "getaddressinfo",                   &getaddressinfo,                {"address"} },
```
> I don't think getaddressinfo should be deprecated. It's not an account rpc

Good catch. getaddressinfo should remain where it was above "getbalance". It's my fault for accidentally moving it into this section during a rebase.

jnewbery commented:
@laanwj - it seems like you're struggling to find time to maintain this PR.

I'd like to push this forwards and make sure it gets in for v0.17. Would you object if I took ownership?

GSPP commented:
Maybe there should be a call returning all addresses and all labels (a bulk call). If you want to dump all address-label pairs this would involve many RPC calls right now. This bulk call should even (optionally?) include internal addresses such as pool addresses. Simply *all* data. I have had the need many times to reflect over absolutely all wallet contents and such an API call would have been very welcome.

jnewbery commented:
I've opened #12892 to supersede this PR. Thanks for all your work @laanwj !

MarcoFalke commented:
Indeed thanks to @laanwj  for initiating it and coming up with the original version of the code. And thx to @ryanofsky  and @jnewbery for picking it up.
